src/site_apportionment/site_apportionment.py

```diff
@@ -0,0 +1,219 @@
+import pandas as pd
+from pandas.api.types import is_numeric_dtype
+import os
+import logging
+
+SitesApportionmentLogger = logging.getLogger(__name__)
+
+# Colunm names redefined for convenience
+ref = "reference"
+ins = "instance"
+period = "period"
+form = "formtype"
+postcode = "postcodes_harmonised"
+percent = "602"
+product = "201"
+pg_num = "pg_numeric"
+civdef = "200"
+
+# Long and short form codes
+short_code = "0006"
+long_code = "0001"
+
+
+def count_unique_codes_in_col(df: pd.DataFrame, code: str) -> pd.DataFrame:
+    """Calculates the number of unique non-empty codes in a column.
+
+    Args:
+        df (pd.DataFrame): A dataframe containing all data
+        code (str): Name of the column containing codes
+
+    Returns:
+        (pd.DataFrame): A copy of originl dataframe with an additional column
+        called the same as code with suffix "_count" countaining the number of
+        unique non-empty codes
+    """
+
+    dfa = df.copy()
+
+    # Select columns that we need
+    cols_need = [ref, period, code]
+    dfa = dfa[cols_need]
+    dfa = dfa[dfa[code].str.len() > 0]
+    dfa.drop_duplicates(inplace=True)
+    dfb = dfa.groupby([ref, period]).agg("count").reset_index()
+    dfb.rename({code: code + "_count"}, axis='columns', inplace=True)
+    df = df.merge(
+        dfb,
+        on=[ref, period],
+        how="left")
+    return df
+
+
+def weights(df):
+    """Calculates site weights based on the percents. Copies the precent value
```

Comment:
Great docstring but would benefit from a summary sentence at the start. 

Also note there's a typo in the word "containing".

Finally, would be helpful to specify what happens if the `percent`, `ref`, `period`, or `ins` column doesn't exist in the dataframe, and they should be listed as parameters I think, with an explanation of each.